File: today/binary-tree-inorder-traversal.ts
/**
 * https://leetcode.com/problems/binary-tree-inorder-traversal/description/
 */
export { inorderTraversal, TreeNode };

function inorderTraversal(root: TreeNode | null): number[] {
  const result: number[] = [];

  function dfs(node: TreeNode | null) {
    if (!node) return;

    if (node.left) dfs(node.left);
    result.push(node.val);
    if (node.right) dfs(node.right);
  }

  dfs(root);
  return result;
}

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

File: returned/python/today7/binary_tree_inorder_traversal.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/description/
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        result = []

        def dfs(node: TreeNode):
            if node.left:
                dfs(node.left)
            
            result.append(node.val)

            if node.right:
                dfs(node.right)
        
        dfs(root)
        return result
